### PR TITLE
gnustep-make, gnustep-base, gnustep-gui, gnustep-back: new package

### PR DESCRIPTION
**Pull request by qmfrederik**
This adds GNUstep to the MINGW64 and UCRT64 MSYS2 environments.  It includes the following packages:

- [gnustep-make](https://repology.org/project/gnustep-make/versions), the build system used by GNUstep
- [gnustep-base](https://repology.org/project/gnustep-base/versions), the base GNUstep library
- [gnustep-gui](https://repology.org/project/gnustep-gui/versions), the GNUstep GUI front-end
- [gnustep-back](https://repology.org/project/gnustep-back/versions), the GNUstep GUI back-end

The GNUstep build system does not allow out-of-dir builds, so I've used `rsync` to clone the GNUstep sources to the environment-specific directory.

### CODE REVIEW

**Comment by qmfrederik** [comment]
Looks like the build is failing because libobjc2 needs to propagate to the package repos first.

**Comment by mmuetzel** [comment]
You'll probably need to wait until the dependencies are built in staging. You can check the queue for when they are ready:
https://packages.msys2.org/queue

File: mingw-w64-gnustep-back/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+_realname=gnustep-back
+pkgbase=mingw-w64-${_realname}
+pkgname=("${MINGW_PACKAGE_PREFIX}-${_realname}")
+pkgver=0.30.0
+pkgrel=1
+pkgdesc="GNUstep GUI Backend (mingw-w64)"
+arch=('any')
+mingw_arch=('mingw64' 'ucrt64')
+url='https://gnustep.github.org/'
+license=('LGPL-2.1-or-later')
```
Comment (Biswa96) [inline mingw-w64-gnustep-back/PKGBUILD:12]:
```suggestion
license=('spdx:LGPL-2.1-or-later')
```

File: mingw-w64-gnustep-back/PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+makedepends=("${MINGW_PACKAGE_PREFIX}-autotools"
+             "${MINGW_PACKAGE_PREFIX}-clang"
+             "${MINGW_PACKAGE_PREFIX}-lld"
+             "${MINGW_PACKAGE_PREFIX}-gnustep-make"
+             rsync)
+depends=("${MINGW_PACKAGE_PREFIX}-gnustep-gui"
+         "${MINGW_PACKAGE_PREFIX}-cairo")
+source=("https://github.com/gnustep/libs-back/releases/download/back-0_30_0/${_realname}-${pkgver}.tar.gz")
```
Comment (Biswa96) [inline mingw-w64-gnustep-back/PKGBUILD:20]:
```suggestion
source=("https://github.com/gnustep/libs-back/releases/download/back-${pkgver//./_}/${_realname}-${pkgver}.tar.gz")
```

File: mingw-w64-gnustep-back/PKGBUILD [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+prepare() {
+  cd "${srcdir}/${_realname}-${pkgver}"
+}
+
+build() {
+  rsync --recursive --times --links "${srcdir}/${_realname}-${pkgver}"/* "${srcdir}/build-${MSYSTEM}"
+  cd "${srcdir}/build-${MSYSTEM}"
+
+  LDFLAGS="-fuse-ld=lld -lstdc++ -lgcc_s" \
```
Comment (Biswa96) [inline mingw-w64-gnustep-back/PKGBUILD:31]: What would happen without these LDFLAGS?

File: mingw-w64-gnustep-base/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+*.patch
```
Comment (Biswa96) [inline mingw-w64-gnustep-base/.gitignore:1]: .gitignore is not required.

File: mingw-w64-gnustep-back/PKGBUILD [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+pkgver=0.30.0
+pkgrel=1
+pkgdesc="GNUstep GUI Backend (mingw-w64)"
+arch=('any')
+mingw_arch=('mingw64' 'ucrt64')
+url='https://gnustep.github.org/'
+license=('LGPL-2.1-or-later')
+makedepends=("${MINGW_PACKAGE_PREFIX}-autotools"
+             "${MINGW_PACKAGE_PREFIX}-clang"
+             "${MINGW_PACKAGE_PREFIX}-lld"
```
[the remark this answers is not in the document]
Comment (qmfrederik) [reply mingw-w64-gnustep-back/PKGBUILD:15]: I've removed it but that creates build failures on MINGW64, where clang does not depend on it: https://packages.msys2.org/package/mingw-w64-x86_64-clang?repo=mingw64

File: mingw-w64-gnustep-back/PKGBUILD (hunk 1 above)
Comment (qmfrederik) [reply mingw-w64-gnustep-back/PKGBUILD]: Should be OK now, thanks!

File: mingw-w64-gnustep-back/PKGBUILD (hunk 2 above)
Comment (qmfrederik) [reply mingw-w64-gnustep-back/PKGBUILD]: Nice, thanks!  Made the change across all PKGBUILD files.

File: mingw-w64-gnustep-back/PKGBUILD (hunk 3 above)
Comment (qmfrederik) [reply mingw-w64-gnustep-back/PKGBUILD:31]: Without this, you'd use the `ld` linker, which does not handle Objective C libraries very well.  Having to specify `-lstdc++ -lgcc_s` appeared to be the result of using `lld`.

File: mingw-w64-gnustep-base/.gitignore (hunk 4 above)
Comment (qmfrederik) [reply mingw-w64-gnustep-base/.gitignore:1]: I found it useful; if not all the `.patch` files which are being downloaded showed up in my `git status` output.  Let me know if you want me to remove it, though.

File: mingw-w64-gnustep-back/PKGBUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (Biswa96) [reply mingw-w64-gnustep-back/PKGBUILD:15]: Oh, I did not noticed it. My apology.

Related question. Would it be possible to add clang64 also?

File: mingw-w64-gnustep-base/.gitignore (hunk 4 above)
Comment (Biswa96) [reply mingw-w64-gnustep-base/.gitignore:1]: Please remove it.

File: mingw-w64-gnustep-back/PKGBUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (qmfrederik) [reply mingw-w64-gnustep-back/PKGBUILD:15]: I've tried to add clang64 but there are a number of weird build errors I would need to get to the bottom to before I can make that work.  It looks like clang somehow gets the wrong include directories for the C runtime when compiling Objective C++ source code, and it gets downhill from there.  That prevented me from adding clang64 support for libobjc2 in the initial PR.

If you don't object, I'd like to get libobjc2 and GNUstep for mingw64 and ucrt64 in first, and then revisit for clang64?

File: mingw-w64-gnustep-base/.gitignore (hunk 4 above)
Comment (qmfrederik) [reply mingw-w64-gnustep-base/.gitignore]: ok - removed.

File: mingw-w64-gnustep-back/PKGBUILD (hunk 5 above)
[the remark this answers is not in the document]
Comment (Biswa96) [reply mingw-w64-gnustep-back/PKGBUILD:15]: It's ok. Just asking.

**Comment by qmfrederik** [comment]
@Biswa96 & others, CI looks green, let me know if there's anything else you'd need from my side.

**Review by Biswa96 — APPROVED** [review]

**Comment by qmfrederik** [comment]
@Biswa96 To add some more context to your earlier questions:
- Support for the clang64 environments requires libobjc2 to support clang64.  Besides CMake getting confused with the C++ headers, there's some work to do to make sure libobjc2's exception handling mechanism works well with clang64.  That's being done here: https://github.com/gnustep/libobjc2/pull/276
- The LDFLAGS environment variable which force linking with stdc++ are required because Objective C has a feature called blocks, and under the hood the libobjc2 runtime will add try/catch statements.  That requires the C++ runtime.

I'll update the libobjc2 and GNUstep ports once support for libc++/clang64 has landed upstream.

**Comment by qmfrederik** [comment]
@Biswa96 @mmuetzel Anything else you need from me to get this merged?

**Comment by qmfrederik** [comment]
@gcasa @triplef @davidchisnall -- fyi, libobjc2 and gnustep-base, gnustep-gui and gnustep-back are now included in msys2.   You all helped to make this happen, thanks!

**Comment by gcasa** [comment]
> @gcasa @triplef @davidchisnall -- fyi, libobjc2 and gnustep-base, gnustep-gui and gnustep-back are now included in msys2.   You all helped to make this happen, thanks!

Thank you as well.  This is absolutely awesome and amazing!!

**Comment by gcasa** [comment]
I think this should be added to the CI for gnustep itself.   Can you guys fully document the build process on msys2 so that this can be reproduced under the GNUstep CI?   Or point us to where it is being built for MSYS2?

**Comment by gcasa** [comment]
Oh, wait... I see it.  Thanks guys... will ask questions if needed.

**Pull request merged**